Add a base Localization class to fluent.runtime

Pike opened the pull request:
This corresponds to what @fluent/bundle does on the js side, but
doesn't try to be async at all.

pmac commented on fluent.runtime/fluent/runtime/fallback.py:
```diff
@@ -0,0 +1,83 @@
+            content = codecs.open(path, 'r', 'utf-8').read()
+            resources.append(self.Resource(content))
+        if resources:
+            yield resources
```
Why not just return the resources list here?

pmac replied:
or instead of making a list at all you can yield each resource as it's created.

Pike replied:
The trick here is that `resources()` can return several different combinations of resource lists, in some setups. I started implementing that, and then optimized to just make that possible. The easiest one (and one we might actually need, depending on deployment), is the `MultiRootResourceLoader`, which might look something like this:

```python
class MultiRootResourceLoader(ResourceLoader):
  def __init__(self, roots):
    self.loaders = [RootedFileResourceLoader(root) for root in roots]
  def resources(self, *args):
    for loader in self.loaders:
      yield from loader.resources(*args)
```

This would allow us to load from both the bedrock and the l10n repo, for example.

In Firefox, we actually allow to mix the different file sources, which makes that code even more complex.

I opted to add that complexity as needed, which is why this code looks complex, but doesn't show why ;-)

pmac replied:
Sure, but the same could be accomplished with something like this:

```python
import itertools

class MultiRootResourceLoader(ResourceLoader):
    def __init__(self, roots):
        self.loaders = [RootedFileResourceLoader(root) for root in roots]

    def resources(self, *args):
        return itertools.chain(*[l.resources(*args) for l in self.loaders])
```

That would still return an iterator that just gave resources one at a time.

pmac commented on fluent.runtime/fluent/runtime/fallback.py:
```diff
@@ -0,0 +1,83 @@
+            yield self._bundle_cache[bundle_pointer]
+            bundle_pointer += 1
+
+    def _iterate_bundles(self):
```
Perhaps I'm missing another layer of this, but it looks to me like the `resource_loader.resources` method should just return an iterator of resources, whereas it's currently returning an iterator containing a single list. So I'd have the `resources()` method of the resource loader just return the list, and then write this like:

```python
    def _iterate_bundles(self):
        for i, locale in enumerate(self.locales):
            bundle = self.bundle_class(self.locales[i:],
                                       functions=self.functions,
                                       use_isolating=self.use_isolating)
            for resource in self.resource_loader.resources(locale, self.resource_ids):
                bundle.add_resource(resource)

            yield bundle
```

Pike replied:
I made the design more obvious here by making `FluentResourceLoader` be both single and multi-root from the get go.

pmac commented on fluent.runtime/fluent/runtime/fallback.py:
```diff
@@ -0,0 +1,83 @@
+            yield resources
+
+    def localize_path(self, path, locale):
+        return path.replace("{locale}", locale)
```
probably better to use `path.format(locale=locale)` here for more robust string parsing.

pmac reviewed (commented):
This looks like it will be quite a bit easier to use :) Thanks! 🥇

pmac commented on fluent.runtime/fluent/runtime/fallback.py:
```diff
@@ -0,0 +1,83 @@
+                continue
+            val, errors = bundle.format_pattern(msg.value, args)
+            return val
+        return id
```
NIT: the `id` variable here masks a Python builtin function with the same name.

pmac commented on fluent.runtime/fluent/runtime/fallback.py:
```diff
@@ -0,0 +1,117 @@
+            for resources in self.resource_loader.resources(locs[0], self.resource_ids):
+                bundle = self._create_bundle(locs)
+                for resource in resources:
+                    bundle.add_resource(resource)
```
Should add the ability to pass in the `allow_override` option to `add_resource` here. Perhaps via a class property like some other options.

Pike replied:
`allow_override` only pays off if you do it for individual resources. Passing it to all resources is the same as just reverting the order in the loader or in the constructor.

I used this feature recently, and it was actually sub-classing the default bundle creation behavior. In terms of the code here, `_iterate_bundles` could be overwritten by a method that "prepends" a resource from a different data source, by doing something like

```python
def _iterate_bundles(self):
    for bundle in super(FunkyLocalization, self)._iterate_bundles():
        bundle.add_resource(self.bass_resource, allow_overrides=True)
        yield bundle
```

Pike committed:
Add a base Localization class to fluent.runtime

This corresponds to what @fluent/bundle does on the js side, but
doesn't try to be async at all.

Pike commented:
This is now upstream on the `next-gen-api` branch.
